### ISSUE TITLE
Push coverage report from Travis-CI

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8532]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#8532|
|Type|enhancement|
|Created|2016-06-28 14:46:04Z|
|Branch|https://github.com/twisted/twisted/tree/8532-travis-codecov|

The travis-ci should also send the coverage report.

Now, since travis only runs on Linux, the overall coverage might be not complete, but I hope that it will help with the coverage for the patch

<details><summary>Searchable metadata</summary>

```
trac-id__8532 8532
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__None None
branch__8532_travis_codecov 8532-travis-codecov
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1467125164569778 1467125164569778
changetime__1468287610821633 1468287610821633
version__None None
owner__adiroiban adiroiban

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_1"></a>|@adiroiban removed owner|
|-|-|

Here is the PR https://github.com/twisted/twisted/pull/259

The coverage is now sent as part of tox and the tox run gets all env variables.

The buildslaves will not send the report twice, as the trial builder does not expose all the environment variables.

Once this is merged I will update the buildslaves to no longer send the report in a dedicated step and to expose the required env variables to codecov.

In this way we can have the coverage builders just like any other trial builder

As a drive by, this enabled travis builds on all branches... so that we don't need to create a PR for each branch.

Please review!

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_2"></a>|@rodrigc set owner to @adiroiban|
|-|-|

OK

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban removed owner|
|-|-|

Putting this back for review as Amber was -1 on this change.

Waiting for more feedback :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

I still hope that we can push the coverage from tox.

I understand that using the default `coverage` env might be a problem ... but then can have a 'coverage-codecov' environment which should only be used by travis and buildbot.

.... but I would argue that we should not keep the TOKEN in tox.ini but rather define it via buildbot

in this way, when users are running the normal coverage, the codecov push will fail since there is no coverage.

still waiting for feedback before doing more changes :)

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_5"></a>|@adiroiban commented|
|-|-|

OK. I have done more work on this.

Running just `tox` on your local machine should behave just like before. No coverage is published.

There is a new `*-coverage-pubish-*` hidden environment which is used to publish the coverage.


You can see a successful publish here https://travis-ci.org/twisted/twisted/jobs/143286418

And that the already existent env `tox -r -e py27-coverage-posix` is not triggering a publish. 

https://buildbot.twistedmatrix.com/builders/ubuntu16.04-py2.7-coverage/builds/584/steps/default/logs/stdio


I hope that in the near future we can also update buildbot to use this new environment.

Please give it another try.

Thanks :)

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi Adi, thanks for this. Three things:

- The branch has conflicts, please merge it forward.

- You don't need to pass in the token on Travis, it will automatically authenticate. Could you replace it with a variable, like ```CODECOV_OPTIONS```, so that we could pass in a build ID token on the buildbots too?

- We only need to install codecov on coverage-publish, not coverage.

Please fix these up and merge away.

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|GitHub set status to `closed`|
|-|-|

In changeset 78429014b30b2523dd5d22aa43a144871ded1eee
```
#!CommitTicketReference repository="" revision="78429014b30b2523dd5d22aa43a144871ded1eee"
Merge 8532-travis-codecov: Publish coverage from Travis-CI.


Author: adiroiban
Reviewer: hawkowl
Fixes: #8532
```
